In [3]:
import numpy as np
import pandas as pd

In [4]:
df_red = pd.read_csv("../data/wine quality/winequality-red.csv") # To load red wine dataset
df_white = pd.read_csv("../data/wine quality/winequality-white.csv") # To load white wine dataset

In [5]:
df_red.head()

fixed acidity;"volatile acidity";"citric acid";"residual sugar";"chlorides";"free sulfur dioxide";"total sulfur dioxide";"density";"pH";"sulphates";"alcohol";"quality"
0   7.4;0.7;0;1.9;0.076;11;34;0.9978;3.51;0.56;9.4;5                                                                                                                     
1   7.8;0.88;0;2.6;0.098;25;67;0.9968;3.2;0.68;9.8;5                                                                                                                     
2  7.8;0.76;0.04;2.3;0.092;15;54;0.997;3.26;0.65;...                                                                                                                     
3  11.2;0.28;0.56;1.9;0.075;17;60;0.998;3.16;0.58...                                                                                                                     
4   7.4;0.7;0;1.9;0.076;11;34;0.9978;3.51;0.56;9.4;5

In [6]:
df_white.head()

fixed acidity;"volatile acidity";"citric acid";"residual sugar";"chlorides";"free sulfur dioxide";"total sulfur dioxide";"density";"pH";"sulphates";"alcohol";"quality"
0   7;0.27;0.36;20.7;0.045;45;170;1.001;3;0.45;8.8;6                                                                                                                     
1  6.3;0.3;0.34;1.6;0.049;14;132;0.994;3.3;0.49;9...                                                                                                                     
2  8.1;0.28;0.4;6.9;0.05;30;97;0.9951;3.26;0.44;1...                                                                                                                     
3  7.2;0.23;0.32;8.5;0.058;47;186;0.9956;3.19;0.4...                                                                                                                     
4  7.2;0.23;0.32;8.5;0.058;47;186;0.9956;3.19;0.4...

## Data and columns names are merged with semicolumns so this data needs to be pre-processed by seperating them to different columns

In [7]:
df_red.shape

(1599, 1)

In [8]:
df_white.shape

(4898, 1)

### There are 1599 entries under Red wine dataset and 4898 in white. There is only 1 column in both that needs to be fixed

In [9]:
###
###Function to fix Column name and return column name as an array, though this was not required
###
def FixColumnNames(df_wine):
    df = df_wine
    cols = df.columns[0].split(";")
    columns = []
    for c in cols:
        columns.append(c.strip('"').strip("'"))
    return columns
    

In [10]:
##Get column names from the red wine dataset, as columns names are same for both red and white
columnNames = FixColumnNames(df_red) 
columnNames

['fixed acidity',
 'volatile acidity',
 'citric acid',
 'residual sugar',
 'chlorides',
 'free sulfur dioxide',
 'total sulfur dioxide',
 'density',
 'pH',
 'sulphates',
 'alcohol',
 'quality']

In [11]:
red_as_numpyarray = df_red.to_numpy() ## Get dateset to numpy array
red_as_numpyarray

array([['7.4;0.7;0;1.9;0.076;11;34;0.9978;3.51;0.56;9.4;5'],
       ['7.8;0.88;0;2.6;0.098;25;67;0.9968;3.2;0.68;9.8;5'],
       ['7.8;0.76;0.04;2.3;0.092;15;54;0.997;3.26;0.65;9.8;5'],
       ...,
       ['6.3;0.51;0.13;2.3;0.076;29;40;0.99574;3.42;0.75;11;6'],
       ['5.9;0.645;0.12;2;0.075;32;44;0.99547;3.57;0.71;10.2;5'],
       ['6;0.31;0.47;3.6;0.067;18;42;0.99549;3.39;0.66;11;6']],
      dtype=object)

In [12]:
split_data_red = [row[0].split(';') for row in red_as_numpyarray] # To split each row by semicolon to get data for separate columns

In [13]:
df_red_modified = pd.DataFrame(split_data_red, columns=columnNames)
df_red_modified

fixed acidity volatile acidity citric acid residual sugar chlorides  \
0              7.4              0.7           0            1.9     0.076   
1              7.8             0.88           0            2.6     0.098   
2              7.8             0.76        0.04            2.3     0.092   
3             11.2             0.28        0.56            1.9     0.075   
4              7.4              0.7           0            1.9     0.076   
...            ...              ...         ...            ...       ...   
1594           6.2              0.6        0.08              2      0.09   
1595           5.9             0.55         0.1            2.2     0.062   
1596           6.3             0.51        0.13            2.3     0.076   
1597           5.9            0.645        0.12              2     0.075   
1598             6             0.31        0.47            3.6     0.067   

     free sulfur dioxide total sulfur dioxide  density    pH sulphates  \
0                     11                   34   0.9978  3.51      0.56   
1                     25                   67   0.9968   3.2      0.68   
2                     15                   54    0.997  3.26      0.65   
3                     17                   60    0.998  3.16      0.58   
4                     11                   34   0.9978  3.51      0.56   
...                  ...                  ...      ...   ...       ...   
1594                  32                   44   0.9949  3.45      0.58   
1595                  39                   51  0.99512  3.52      0.76   
1596                  29                   40  0.99574  3.42      0.75   
1597                  32                   44  0.99547  3.57      0.71   
1598                  18                   42  0.99549  3.39      0.66   

     alcohol quality  
0        9.4       5  
1        9.8       5  
2        9.8       5  
3        9.8       6  
4        9.4       5  
...      ...     ...  
1594    10.5       5  
1595    11.2       6  
1596      11       6  
1597    10.2       5  
1598      11       6  

[1599 rows x 12 columns]

In [14]:
white_as_numpyarray = df_white.to_numpy() ## Get dateset to numpy array
white_as_numpyarray

array([['7;0.27;0.36;20.7;0.045;45;170;1.001;3;0.45;8.8;6'],
       ['6.3;0.3;0.34;1.6;0.049;14;132;0.994;3.3;0.49;9.5;6'],
       ['8.1;0.28;0.4;6.9;0.05;30;97;0.9951;3.26;0.44;10.1;6'],
       ...,
       ['6.5;0.24;0.19;1.2;0.041;30;111;0.99254;2.99;0.46;9.4;6'],
       ['5.5;0.29;0.3;1.1;0.022;20;110;0.98869;3.34;0.38;12.8;7'],
       ['6;0.21;0.38;0.8;0.02;22;98;0.98941;3.26;0.32;11.8;6']],
      dtype=object)

In [15]:
split_data_white = [row[0].split(';') for row in white_as_numpyarray] # To split each row by semicolon to get data for separate columns

In [16]:
df_white_modified = pd.DataFrame(split_data_white, columns=columnNames)
df_white_modified

fixed acidity volatile acidity citric acid residual sugar chlorides  \
0                7             0.27        0.36           20.7     0.045   
1              6.3              0.3        0.34            1.6     0.049   
2              8.1             0.28         0.4            6.9      0.05   
3              7.2             0.23        0.32            8.5     0.058   
4              7.2             0.23        0.32            8.5     0.058   
...            ...              ...         ...            ...       ...   
4893           6.2             0.21        0.29            1.6     0.039   
4894           6.6             0.32        0.36              8     0.047   
4895           6.5             0.24        0.19            1.2     0.041   
4896           5.5             0.29         0.3            1.1     0.022   
4897             6             0.21        0.38            0.8      0.02   

     free sulfur dioxide total sulfur dioxide  density    pH sulphates  \
0                     45                  170    1.001     3      0.45   
1                     14                  132    0.994   3.3      0.49   
2                     30                   97   0.9951  3.26      0.44   
3                     47                  186   0.9956  3.19       0.4   
4                     47                  186   0.9956  3.19       0.4   
...                  ...                  ...      ...   ...       ...   
4893                  24                   92  0.99114  3.27       0.5   
4894                  57                  168   0.9949  3.15      0.46   
4895                  30                  111  0.99254  2.99      0.46   
4896                  20                  110  0.98869  3.34      0.38   
4897                  22                   98  0.98941  3.26      0.32   

     alcohol quality  
0        8.8       6  
1        9.5       6  
2       10.1       6  
3        9.9       6  
4        9.9       6  
...      ...     ...  
4893    11.2       6  
4894     9.6       5  
4895     9.4       6  
4896    12.8       7  
4897    11.8       6  

[4898 rows x 12 columns]

In [56]:
df_white_modified.to_csv("../data/PreProcessed/WhiteWineFixed.csv", index=False)
df_red_modified.to_csv("../data/PreProcessed/RedWineFixed.csv", index=False)